src/snapshot.ts

```diff
@@ -0,0 +1,346 @@
+  }
+}
+
+function currency(...items: Item[]): () => number {
+  const unitCost: [Item, number][] = items.map((i) => {
+    const coinmaster = Coinmaster.all().find((c) => sellPrice(c, i) > 0);
+    if (!coinmaster) {
+      throw `Invalid coinmaster item ${i}`;
+    } else {
+      return [i, sellPrice(coinmaster, i)];
+    }
+  });
+  return () => Math.max(...unitCost.map(([item, cost]) => garboValue(item) / cost));
+}
+function complexCandy(): [Item, () => number][] {
+  const candies = Item.all().filter((i) => i.candyType === "complex");
+  const candyLookup: Item[][] = [[], [], [], [], []];
+
+  for (const candy of candies) {
+    const id = toInt(candy) % 5;
+    if (candy.tradeable) {
+      candyLookup[id].push(candy);
+    }
+  }
+  const candyIdPrices: [Item, () => number][] = candies
+    .filter((i) => !i.tradeable)
+    .map((i) => [i, () => Math.min(...candyLookup[toInt(i) % 5].map((i) => garboValue(i)))]);
+  return candyIdPrices;
+}
+
+const specialValueLookup = new Map<Item, () => number>([
+  [
+    $item`Freddy Kruegerand`,
+    currency(...$items`bottle of Bloodweiser, electric Kool-Aid, Dreadsylvanian skeleton key`),
+  ],
+  [$item`Beach Buck`, currency($item`one-day ticket to Spring Break Beach`)],
+  [$item`Coinspiracy`, currency(...$items`Merc Core deployment orders, karma shawarma`)],
+  [$item`FunFunds™`, currency($item`one-day ticket to Dinseylandfill`)],
+  [$item`Volcoino`, currency($item`one-day ticket to That 70s Volcano`)],
+  [$item`Wal-Mart gift certificate`, currency($item`one-day ticket to The Glaciest`)],
+  [$item`Rubee™`, currency($item`FantasyRealm guest pass`)],
+  [$item`Guzzlrbuck`, currency($item`Never Don't Stop Not Striving`)],
+  ...complexCandy(),
+  [
+    $item`Merc Core deployment orders`,
+    () => garboValue($item`one-day ticket to Conspiracy Island`),
+  ],
+]);
+
+function printSnapshot(snapshot: Snapshot): void {
+  const value = snapshot.value(garboValue);
+  const printProfit = (details: ItemDetail[]) => {
+    for (const { item, quantity, value } of details) {
+      print(`  ${item} (${quantity}) @ ${Math.floor(value)}`);
+    }
+  };
+  const lowValue = value.itemDetails.filter((detail) => detail.value < 0);
+  const highValue = value.itemDetails.filter((detail) => detail.value > 0);
+
+  print(`Total Snapshot Value: ${value.total}`);
+  print(`Of that, ${value.meat} came from meat and ${value.items} came from items`);
+  print(` You gained meat on ${highValue.length} items including:`);
+  printProfit(highValue);
+  print(` You lost meat on ${lowValue.length} items including:`);
+  printProfit(lowValue);
+}
+
+const garboValueCache = new Map<Item, number>();
+export function garboValue(item: Item): number {
+  const cachedValue = garboValueCache.get(item);
+  if (cachedValue === undefined) {
+    const specialValueCompute = specialValueLookup.get(item);
+    const value = specialValueCompute ? specialValueCompute() : getSaleValue(item);
```

Comment:
given that getSaleValue also caches, it feels a little redundant to do this. Should we entire bypass getSaleValue and recreate the code, caching it directly? We'll be building up _two_ maps for every item we calculate the price of.

Reply:
I cached on this later because we bypass getSaleVslue for many items, and I was going to only cache those but it seemed silly to cache on just that later when I could just cache the full flow instead. I am not even a little worried about the memory implications of this